main.tf

```diff
@@ -18,16 +18,35 @@ locals {
   prefix       = var.prefix == "" ? "" : "${var.prefix}-"
   folders_list = [for name in var.names : try(google_folder.folders[name], "")]
   first_folder = try(local.folders_list[0], {})
-  folder_admin_roles_map_data = merge([
-    for name, config in var.per_folder_admins : {
-      for role in config.roles != null ? config.roles : var.folder_admin_roles : "${name}-${role}" =>
-      {
+
+  # Handle roles for per_folder_admins if provided
+  folder_admin_roles_map_data = flatten([
+    for name, config in var.per_folder_admins : [
+      for role in config.roles != null ? config.roles : var.folder_admin_roles : {
         name    = name,
         role    = role,
-        members = config.members,
+        members = config.members
       }
-    }
-  ]...)
+    ]
+  ])
+
+  # Handle roles for all_folder_admins if provided, applied to all folders only if they are not part of per_folder_admins
+  folder_admin_roles_all_folders = flatten([
+    for folder in var.names : [
+      for role in var.folder_admin_roles : {
+        name    = folder,
+        role    = role,
+        members = var.all_folder_admins
+      }
+    ]
+    # Only add roles for all_folder_admins if they are not already present in per_folder_admins
+    if length(var.all_folder_admins) > 0 && !contains(keys(var.per_folder_admins), folder)
```

Comment:
You might also consider running `distinct()` on the final list as a simpler solution to eliminating duplicates.